pkg/controller/backup_controller.go

```diff
@@ -595,60 +629,66 @@ func recordBackupMetrics(log logrus.FieldLogger, backup *velerov1api.Backup, bac
 	serverMetrics.RegisterVolumeSnapshotFailures(backupScheduleName, backup.Status.VolumeSnapshotsAttempted-backup.Status.VolumeSnapshotsCompleted)
 }
 
-func persistBackup(backup *pkgbackup.Request, backupContents, backupLog *os.File, backupStore persistence.BackupStore, log logrus.FieldLogger) []error {
+func persistBackup(backup *pkgbackup.Request,
+	backupContents, backupLog *os.File,
+	backupStore persistence.BackupStore,
+	log logrus.FieldLogger,
+	csiVolumeSnapshots []*snapshotv1beta1api.VolumeSnapshot,
+	volumeSnapshotContents []*snapshotv1beta1api.VolumeSnapshotContent,
+) []error {
 	errs := []error{}
 	backupJSON := new(bytes.Buffer)
 
 	if err := encode.EncodeTo(backup.Backup, "json", backupJSON); err != nil {
 		errs = append(errs, errors.Wrap(err, "error encoding backup"))
 	}
 
-	volumeSnapshots := new(bytes.Buffer)
-	gzw := gzip.NewWriter(volumeSnapshots)
-
-	if err := json.NewEncoder(gzw).Encode(backup.VolumeSnapshots); err != nil {
-		errs = append(errs, errors.Wrap(err, "error encoding list of volume snapshots"))
-	}
-	if err := gzw.Close(); err != nil {
-		errs = append(errs, errors.Wrap(err, "error closing gzip writer"))
+	// Velero-native volume snapshots (as opposed to CSI ones)
+	nativeVolumeSnapshots, err := encodeToJSONGzip(backup.VolumeSnapshots)
+	if err != nil {
+		errs = append(errs, err)
```

Comment:
is this change required?
I feel the previous version was more informative. WDYT?

Reply:
You mean using the helper function?

I don't think it's strictly required, no. It was mostly based on a review comment to reduce repetition in the function.

Reply:
Not the helper function, but the error wrapping part.

Reply:
Oh, I see - so the error wrapping should be the same, just split between the function and up here, but the one thing we are losing is the information about exactly what's being encoded, which would definitely make debugging harder.

I had considered adding a string for the error message as input to the function, so it would look like `encodeToJSONGzip(backup.VolumeSnapshots, "list of volumesnapshots")`, and the error would be more descriptive when returned.

Would that help?

Reply:
For example, L649 here should be equivalent to L610 in the previous code.

Reply:
I've added a second argument to help produce more contextual errors, so hopefully they're less confusing when they happen.

Reply:
ah. That's correct.  My bad 🤦